toolkit/components/uniffi-bindgen-gecko-js/src/templates/js/Object.jsm

```diff
@@ -33,9 +35,36 @@ class {{ object.nm() }} {
 
     {%- for meth in object.methods() %}
     {{ meth.nm() }}({{ meth.arg_names() }}) {
-        {% call js::call_method(meth, type_) %}
+        return this.runMethod(() => {
+            {% call js::call_method(meth, type_, object) %}
+        });
     }
     {%- endfor %}
+
+    destroy() {
+        this.destroyed = true;
+        // If the call counter is not zero, there are ongoing calls that haven't concluded
+        // yet. The function calls themselves will make sure to deallocate the object once the last
+        // one concludes and we will prevent any new calls by throwing a UniFFIError
+        if (this.callCounter === 0) {
+            {{ ci.scaffolding_name() }}.{{ object.ffi_object_free().nm() }}(this.ptr);
+        }
+    }
+
+    runMethod(callback) {
```

Comment:
Nit: maybe this should be something like `uniffiRunMethod` just to avoid any chance of a name conflict with user methods.